[NFC][RemoteInspection] Add an opaque AddressSpace field to RemoteAddress

augusto2112 opened the pull request:
Add an extra opaque field to AddressSpace, which can be used by clients
of RemoteInspection to distinguish between different address spaces.

LLDB employs an optimization where it reads memory from files instead of
the running process whenever it can to speed up memory reads (these can
be slow when debugging something over a network). To do this, it needs
to keep track whether an address originated from a process or a file. It
currently distinguishes addresses by setting an unused high bit on the
address, but because of pointer authentication this is not a reliable
solution. In order to keep this optimization working, this patch adds an
extra opaque AddressSpace field to RemoteAddress, which LLDB can use on
its own implementation of MemoryReader to distinguish between addresses.

This patch is NFC for the other RemoteInspection clients, as it adds
extra information to RemoteAddress, which is entirely optional and if
unused should not change the behavior of the library.

Although this patch is quite big the changes are largely mechanical,
replacing threading StoredPointer with RemoteAddress.

rdar://148361743

augusto2112 commented:
https://github.com/swiftlang/llvm-project/pull/10928
@swift-ci test

adrian-prantl commented on include/swift/Remote/RemoteAddress.h:
```diff
@@ -18,85 +18,274 @@
-  uint64_t Data;
+  uint64_t Data = 0;
+  // Buffer to be used by subclasses
+  uint64_t Buffer = 0;
```
OpaqueStorage?

adrian-prantl commented on include/swift/Remote/RemoteAddress.h:
```diff
@@ -18,85 +18,274 @@
 
-  explicit operator bool() const {
-    return Data != 0;
+  static RemoteAddress makeAddressFromPtr(const void *localPtr) {
```
If possible, can we get rid of this constructor?

adrian-prantl commented on include/swift/Remote/RemoteAddress.h:
```diff
@@ -18,85 +18,274 @@
+  virtual ~RemoteAddress() = default;
+
+  template <class IntegerType>
+  RemoteAddress fromThis(IntegerType data) const {
```
Instead of this how about

`RemoteAddress(value, otherRemoteAddress.getAddressSpace())`

augusto2112 commented:
https://github.com/swiftlang/llvm-project/pull/10928
@swift-ci test

mikeash commented on include/swift/Basic/RelativePointer.h:
```diff
@@ -132,18 +132,21 @@
 /// Apply a relative offset to a base pointer. The offset is applied to the base
 /// pointer using sign-extended, wrapping arithmetic.
-template<typename BasePtrTy, typename Offset>
+template <typename BasePtrTy, typename Offset>
```
Maybe it had too much coffee this morning.

augusto2112 commented:
https://github.com/swiftlang/llvm-project/pull/10928
@swift-ci test

augusto2112 committed:
[NFC][RemoteInspection] Add an opaque AddressSpace field to RemoteAddress

Add an extra opaque field to AddressSpace, which can be used by clients
of RemoteInspection to distinguish between different address spaces.

LLDB employs an optimization where it reads memory from files instead of
the running process whenever it can to speed up memory reads (these can
be slow when debugging something over a network). To do this, it needs
to keep track whether an address originated from a process or a file. It
currently distinguishes addresses by setting an unused high bit on the
address, but because of pointer authentication this is not a reliable
solution. In order to keep this optimization working, this patch adds an
extra opaque AddressSpace field to RemoteAddress, which LLDB can use on
its own implementation of MemoryReader to distinguish between addresses.

This patch is NFC for the other RemoteInspection clients, as it adds
extra information to RemoteAddress, which is entirely optional and if
unused should not change the behavior of the library.

Although this patch is quite big the changes are largely mechanical,
replacing threading StoredPointer with RemoteAddress.

rdar://148361743

augusto2112 commented:
https://github.com/swiftlang/llvm-project/pull/10928
@swift-ci test

MaxDesiatov commented:
Could this PR have broken the build here? https://ci.swift.org/job/swift-PR-Linux-smoke-test/21185/console

```
16:25:11  /home/build-user/llvm-project/lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:175:36: error: no member named 'getAddressData' in 'swift::remote::RemoteRef<swift::reflection::TargetFieldDescriptor<swift::InProcess>>'
16:25:11      auto offset = field_descriptor.getAddressData() -
16:25:11                    ~~~~~~~~~~~~~~~~ ^
16:25:11  /home/build-user/llvm-project/lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:176:52: error: no member named 'getAddressData' in 'swift::remote::RemoteRef<void>'
16:25:11                    field_descriptors.startAddress().getAddressData();
16:25:11                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
16:25:11  2 errors generated.
16:25:11  [502/1131][ 44%][86.096s] B
```

augusto2112 commented:
> @augusto2112 Could this PR have broken the build here? https://ci.swift.org/job/swift-PR-Linux-smoke-test/21185/console
> 
> ```
> 16:25:11  /home/build-user/llvm-project/lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:175:36: error: no member named 'getAddressData' in 'swift::remote::RemoteRef<swift::reflection::TargetFieldDescriptor<swift::InProcess>>'
> 16:25:11      auto offset = field_descriptor.getAddressData() -
> 16:25:11                    ~~~~~~~~~~~~~~~~ ^
> 16:25:11  /home/build-user/llvm-project/lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:176:52: error: no member named 'getAddressData' in 'swift::remote::RemoteRef<void>'
> 16:25:11                    field_descriptors.startAddress().getAddressData();
> 16:25:11                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
> 16:25:11  2 errors generated.
> 16:25:11  [502/1131][ 44%][86.096s] B
> ```

There's an accompanying LLDB PR (https://github.com/swiftlang/llvm-project/pull/10928). The CI job probably picked up this commit but not the LLDB commit. It just started a new job which I assume should be fine now.
